mhkit/acoustics/analysis.py

```diff
@@ -0,0 +1,806 @@
+            warnings.warn(
+                f"'spsd' does not have 'freq' as a dimension and has multiple dimensions. "
+                f"Using the first dimension '{spsd.dims[0]}' as 'freq'."
+            )
+        # Assign the 0th dimension as 'freq'
+        spsd = spsd.rename({spsd.dims[0]: "freq"})
+
+    # Ensure 'freq' dimension exists in 'sensitivity_curve'
+    if "freq" not in sensitivity_curve.dims:
+        if len(sensitivity_curve.dims) > 1:
+            # Issue a warning and assign the 0th dimension as 'freq'
+            warnings.warn(
+                f"'sensitivity_curve' does not have 'freq' as a dimension \
+                      and has multiple dimensions. "
+                f"Using the first dimension '{sensitivity_curve.dims[0]}' as 'freq'."
+            )
+        # Assign the 0th dimension as 'freq'
+        sensitivity_curve = sensitivity_curve.rename(
+            {sensitivity_curve.dims[0]: "freq"}
+        )
+
+    # Create a copy of spsd to avoid in-place modification
+    spsd_calibrated = spsd.copy()
+
+    # Read calibration curve
+    freq = sensitivity_curve.dims[0]
+    # Interpolate calibration curve to desired value
+    calibration = sensitivity_curve.interp(
+        {freq: spsd_calibrated["freq"]}, method="linear"
+    ).drop_vars(freq)
+    # Fill missing with provided value
+    calibration = calibration.fillna(fill_value)
+
+    # Subtract from sound pressure spectral density
+    sensitivity_ratio = 10 ** (calibration / 10)  # V^2/uPa^2
+    spsd_calibrated /= sensitivity_ratio  # uPa^2/Hz
+    spsd_calibrated /= 1e12  # Pa^2/Hz
+    spsd_calibrated.attrs["units"] = "Pa^2/Hz"
+
+    return spsd_calibrated
+
+
+def sound_pressure_spectral_density_level(spsd: xr.DataArray) -> xr.DataArray:
+    """
+    Calculates the sound pressure spectral density level from
+    the mean square sound pressure spectral density.
+
+    Parameters
+    ----------
+    spsd: xarray.DataArray (time, freq)
+        Mean square sound pressure spectral density in Pa^2/Hz
+
+    Returns
+    -------
+    spsdl: xarray.DataArray (time, freq)
+        Sound pressure spectral density level [dB re 1 uPa^2/Hz]
+        indexed by time and frequency
+    """
+
+    # Reference value of sound pressure
+    reference = 1e-12  # Pa^2 to 1 uPa^2
+
+    # Sound pressure spectral density level from mean square values
+    lpf = 10 * np.log10(spsd.values / reference)
+
+    spsdl = xr.DataArray(
+        lpf.astype(np.float32),
+        coords={"time": spsd["time"], "freq": spsd["freq"]},
+        attrs={
+            "units": "dB re 1 uPa^2/Hz",
+            "long_name": "Sound Pressure Spectral Density Level",
+        },
+    )
+
+    return spsdl
+
+
+def _validate_method(
+    method: Union[str, Dict[str, Union[float, int]]]
+) -> Tuple[str, Optional[Union[float, int]]]:
+    """
+    Validates the 'method' parameter and returns the method name and argument (if any).
```

Comment:
Reading through this the docstring could be improved by describing what this function raises and  showing examples:

```